### PR TITLE
Thread affinity and detach.

### PR DESCRIPTION
**Pull request by kpamnany**
Modified version of ~~https://github.com/libuv/libuv/pull/280~~ https://github.com/libuv/libuv/pull/597. Adds 3 calls to set thread affinity, get thread affinity, and detach a thread. Windows support is coded but untested!

```
int uv_thread_setaffinity(uv_thread_t *tid,
                          char *cpumask,
                          char *oldmask,
                          size_t mask_size);
```

Sets the specified thread's affinity to `cpumask`, which must be specified in bytes. Returns the previous affinity setting in `oldmask`, if provided. On Unix, uses `pthread_getaffinity_np()` to get the affinity setting and maps the `cpu_set_t` to bytes in `oldmask`. Then maps the bytes in `cpumask` to a `cpu_set_t` and uses `pthread_setaffinity_np()`. On Windows, maps the bytes in `cpumask` to a bitmask and uses `SetThreadAffinityMask()` which returns the previous affinity setting.

```
int uv_thread_getaffinity(uv_thread_t *tid,
                          char *cpumask,
                          size_t mask_size);
```

Gets the specified thread's affinity setting. On Unix, maps the `cpu_set_t` returned by `pthread_getaffinity_np()` to bytes in `cpumask`. On Windows, uses `SetThreadAffinityMask()` with the process affinity mask to get the thread's old affinity, and then again, to reset it.

```
int uv_thread_detach(uv_thread_t *tid);
```

Detaches the specified thread so it will be cleaned up on exit automatically; joining it is no longer necessary or possible. Uses `pthread_detach()` on Unix and `CloseHandle()` on Windows.

### CODE REVIEW

**Comment by tkelman** [comment]
automake takes about 5 minutes to compile a local version of, would make the diff a lot less noisy

**Comment by kpamnany** [comment]
Not really much less noisy, but okay.

**Comment by tkelman** [comment]
Looks nicer to me, thanks. I think we could make a test branch of Julia and run appveyor on this if it would help.

**Comment by kpamnany** [comment]
If that's the best way to get this tested on Windows, then sure. In any case, I'll work on building Julia in Windows per the instructions in `README.windows`.

**Comment by tkelman** [comment]
This looks okay for me in local builds, both win32 and win64, so if you and/or upstream libuv are happy with it I think we could just merge (unless anyone has any substantive feedback). I'm guessing that stock Julia's tests won't really execute the new features here?

**Comment by kpamnany** [comment]
Thanks for checking! No word on the PR in upstream libuv, though the API is based on input from one of the libuv devs.

This is for use by the `threading` branch, to abstract away platform specifics. There is no specific test case for thread affinitization as wrong affinities will only impact performance, not correctness.

FYI, we'll need to add something for atomics as well, soon.

Any comments? @JeffBezanson, @ArchRobison, @vtjnash?

File: src/uv-common.c [hunk 1]
```diff
@@ -295,6 +295,111 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+  if (!GetProcessAffinityMask(hproc, &pam, &sam))
+    return uv_translate_sys_error(GetLastError());
+
+  tam = SetThreadAffinityMask(*tid, pam);
+  if (!tam)
+    return uv_translate_sys_error(GetLastError());
+  SetThreadAffinityMask(*tid, tam);
+
+  for (i = 0;  i < UV_CPU_SETSIZE;  i++)
+    cpumask[i] = (tam & (1 << i));
```
Comment (ArchRobison) [inline src/uv-common.c]: If i>=8, then this line will unconditionally set cpumask[i] to 0.  Did you mean cpumask[i] = (tam>>i & 1)?

File: src/uv-common.c [hunk 2]
```diff
@@ -295,6 +295,111 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+    return uv_translate_sys_error(GetLastError());
+
+  if (oldmask) {
+    for (i = 0;  i < UV_CPU_SETSIZE;  i++)
+      oldmask[i] = (oldtam & (1 << i));
+  }
+
+  return 0;
+#else
+  int i, r;
```
Comment (ArchRobison) [inline src/uv-common.c:336]: I recommend moving these declarations to their points of initialization.  That way, their live range is clearer to maintainers.  Of course doing this for i requires C99.  I don't know if this package uses C99 anyway or not.

**Comment by ArchRobison** [comment]
Where's the documentation for these new entry points?

File: src/uv-common.c (hunk 2 above)
Comment (tkelman) [inline src/uv-common.c:336]: libuv generally wants things to be compile-able by oldish msvc versions that aren't c99 compliant, though for parts that are under this branch of `#ifdef _WIN32` they might be more lenient?

File: src/uv-common.c (hunk 1 above)
Comment (kpamnany) [inline src/uv-common.c]: I meant `cpumask[i] = (tam & (1 << i)) ? 1 : 0` but your way is better. Thank you.

File: src/uv-common.c (hunk 2 above)
Comment (kpamnany) [inline src/uv-common.c:336]: From https://github.com/libuv/libuv/blob/master/CONTRIBUTING.md:
- Use C89-compliant syntax. In other words, variables can only be declared at the top of a scope (function, if/for/while-block).

**Comment by kpamnany** [comment]
Where should I document them? Upstream libuv has `docs/` and I will update the PR there to add these calls, but Julia's libuv appears to have been forked before documentation was added.

**Comment by tkelman** [comment]
I'd almost say inline comments, or in this thread, for our fork. Dunno.

And just to be clear, by "This looks okay for me in local builds, both win32 and win64" I meant Julia built with this branch passes its tests. The libuv tests won't build on our fork, but this is a pre-existing problem unrelated to this PR. And I haven't tested with MSVC.

**Comment by kpamnany** [comment]
Added the documentation I'd written in the upstream libuv PR to the top of this thread.

File: src/uv-common.c (hunk 2 above)
Comment (ArchRobison) [inline src/uv-common.c:336]: Thanks for the citation.  Sigh.

**Comment by tkelman** [comment]
@vtjnash ok to merge this?

**Pull request merged**